File: leetcode/20. Valid Parentheses.ts
//https://leetcode.com/problems/valid-parentheses/description/
function isValid(s: string): boolean {
    const chars = {
        ')':'(',
        '}':'{',
        ']':'['
    }

    const str: string[] = s.split('')
    //keep stack of all the elements
    const stack: string[] = []
    for(let i = 0; i < str.length; i++) {
        const curr = str[i]
        //check if it's a closing tag
        if(curr in chars) {
            const lastIn = stack.pop()
            //check that it's the correct closing tag aka check the last in from stack and the current element (that gets its opening tag from dict)
            if(chars[curr] !== lastIn) {
                return false
            }
        } else {
            //if it's not a closing tag then push into stack
            stack.push(curr)
        }
    }
    //make sure there are no elements to check
    if(stack.length > 0) return false

    return true
};

File: leetcode/20. Valid Parentheses.go
// https://leetcode.com/problems/valid-parentheses/description/
func isValid(s string) bool {
	if len(s) == 0 || len(s)%2 == 1 {
		return false
	}

	pairs := map[rune]rune{
		'(': ')',
		'{': '}',
		'[': ']',
	}
	stack := []rune{}

	for _, r := range s {
		if _, ok := pairs[r]; ok {
			stack = append(stack, r)
		} else if len(stack) == 0 || pairs[stack[len(stack)-1]] != r {
			return false
		} else {
			stack = stack[:len(stack)-1]
		}
	}

	return len(stack) == 0
}

func isValid2(s string) bool {
	if len(s) == 0 || len(s)%2 == 1 {
		return false
	}

	m := map[rune]rune{
		')': '(',
		'}': '{',
		']': '[',
	}

	stack := []rune{}

	for _, c := range s {
		pair, ok := m[c]
		if !ok {
			stack = append(stack, c)
			continue
		}

		if len(stack) == 0 {
			return false
		}

		if pair != stack[len(stack)-1] {
			return false
		}

		// pop the element at the end (we compare above, this is for better memory usage)
		stack = stack[:len(stack)-1]
	}

	return len(stack) == 0
}
